onnxruntime/core/graph/graph_viewer.cc

```diff
@@ -36,11 +36,11 @@ struct PriorityNodeCompare {
     }
 
     // nodes of forward pass will be output first
-    auto n1_attrs = n1->GetAttributes();
-    auto n2_attrs = n2->GetAttributes();
-    int64_t n1_is_forward = static_cast<int64_t>(n1_attrs.find(kBackwardNodeAttributeName) == n1_attrs.cend()) ||
+    auto const& n1_attrs = n1->GetAttributes();
+    auto const& n2_attrs = n2->GetAttributes();
+    int64_t n1_is_forward = static_cast<int64_t>(n1->isForwardNode()) ||
```

Comment:
This change shaves off nearly 2s from the original 20s of TransformGraph and will be maximum efficient when loading models for inference.

For training models the hash computation is still done once. Before the hash has potentially been computed twice, one time for find and one time for at. By fetching the iterator of std::fine and using it later to get the value of i the number of hash computation could be halved (saving ~1s instead of 2s).

If it's known that the model is used for inference only it'd be great if PriorityNodeCompare could skip this test altogether. This could be achieved most efficient by making this a template class and specialize for inference / training.